connectivity/drivers/ble/FEATURE_BLE/COMPONENT_CYW43XXX/CyH4TransportDriver.cpp

```diff
@@ -194,11 +245,20 @@ uint16_t CyH4TransportDriver::write(uint8_t type, uint16_t len, uint8_t *pData)
 
     while (i < len + 1) {
         uint8_t to_write = i == 0 ? type : pData[i - 1];
+#if defined(CYW43XXX_UNBUFFERED_UART)
+        while (uart.writeable() == 0);
+        uart.write(&to_write, 1);
+#else
         while (cyhal_uart_writable(&uart) == 0);
         cyhal_uart_putc(&uart, to_write);
+#endif
         ++i;
     }
+#if defined(CYW43XXX_UNBUFFERED_UART)
+    while (uart.writeable() == 0);
```

Comment:
This doesn't necessarily do the same thing as the HAL counterpart. At least on PSoC devices, uart.writeable() just  indicate whether there is space in the serial adapter's FIFO to write a character. 

Similarly, putc will block until the character can written to the adapter, but does not wait until it the byte has been sent out over the wire. 

I don't know how other MCUs function, but the Mbed HAL documentation for putc doesn't explicitly state that the data will be sent over the wire before it returns:
> This is a blocking call, waiting for a peripheral to be available  for writing

This is important because deasserting bt_dev_wake before the write to the radio completes can cause communication failures.

The equivalent Mbed HAL-level function to `cyhal_uart_is_tx_active` would be `serial_tx_active`, but I don't know of any way in current Mbed to access this function from the UnbufferedSerial class.

Reply:
Am I correct in supposing that it is important to know that all the bytes have been transmitted before calling  `deassert_bt_dev_wake` ? 
@kjbracey-arm Is there anything we can do to know when the hardware queue is empty ?

Reply:
Correct. Calling `serial_tx_active` should provide that answer; the question is how to access that when all you have is an UnbufferedSerial object.

Reply:
`serial_tx_active` is only available if async is enabled, unfortunate that this is not always enabled and implemented.

I wonder if using the HAL function in this case, here in this component, would cause the trouble (the requirement would be to have async serial available) ?

Reply:
on our platform uart FIFO is disabled by default, https://github.com/ARMmbed/mbed-os/blob/77638f9b725592629b38b9124e7b56f72cc75401/targets/TARGET_STM/serial_api.c#L633 so as far as i can understand `uart.writable()` and `serial_tx_active` have the same meaning, but this is not true for other platforms using the serial hardware FIFO.

Reply:
just for the purpose of trying i've made some changes to be able to use `serial_tx_active`. But something is not working and the application is crashing. I will investigate further on it also running ble integration tests.

Reply:
just discovered that the problem was me, and `serial_tx_active` is working as expected, but i cannot found a way to call `serial_tx_active` hal function directly from `CyH4TransportDriver`. 
Any suggestion to how retrive `serial_t *obj` from there is really appreciated. 
So with https://github.com/ARMmbed/mbed-os/pull/14227/commits/40a3636a1b10a2b9afbb66ba952c5bbc371251dd i've made small changes to `SerialBase` and `UnbufferedUart` to reach `serial_tx_active` function if `DEVICE_SERIAL_ASYNCH` is defined.

Reply:
@0xc0170  @kyle-cypress  is  https://github.com/ARMmbed/mbed-os/commit/4c7d7e7e9a460cf52a45941cc77f3bac252b65dc an acceptable solution ?

Reply:
The changes in the BLE HCI driver look reasonable to me. 
As do the changes in the SerialBase and UnbufferedSerial, but @0xc0170 may have stronger opinions there than I do.

Reply:
> @0xc0170 @kyle-cypress is 4c7d7e7 an acceptable solution ?

Please send  the extension to SerialBase/UnbufferedSerial via new PR (as it's feature update).